### ISSUE TITLE
Deadlock after BrokenProcessPool exception

### ISSUE THREAD

**Issue by jackdent** [noxdafox/pebble#69]
I've been observing the occasional deadlock in pebble's `ProcessPool`, which seem to follow `BrokenProcessPool` exceptions.

I was able to replicate the deadlock with the following snippet:
```
from pebble import ProcessPool
from pebble.pool.base_pool import ERROR

def fn(i):
    print(i)

with ProcessPool() as pool:
    future = pool.map(fn, range(100), timeout=5)
    iterator = future.result()
    pool._context.state = ERROR

    while True:
        try:
            result = next(iterator)
        except StopIteration:
            break
        except Exception as err:
            print(err)

```
I set `pool._context.state = ERROR` to simulate a `BrokenProcessPool` midway through my workload.

If I attach a debugger right before `result = next(iterator)`, it becomes clear that we are waiting on the result of a future which never resolves:
```
(Pdb) where
  /Users/jackd/crash.py(15)<module>()
-> result = next(iterator)
  /usr/local/lib/python3.7/site-packages/pebble/pool/base_pool.py(206)next()
-> result = next(self._results)
  /usr/local/lib/python3.7/site-packages/pebble/pool/base_pool.py(200)<genexpr>()
-> self._results = chain.from_iterable(chunk_result(f) for f in futures)
  /usr/local/lib/python3.7/site-packages/pebble/pool/base_pool.py(232)chunk_result()
-> return future.result()
  /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/_base.py(430)result()
-> self._condition.wait(timeout)
  /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py(296)wait()
-> waiter.acquire()
> /usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py(186)sigint_handler()
-> def sigint_handler(self, signum, frame):
```

**Comment by sankarravi** [comment]
For anyone else who stumbles onto this issue, @jackdent and I ended up working around this using Python 3's native [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html) and the smaller [`timeout-decorator`](https://github.com/pnpnpn/timeout-decorator) library for timing out processes. For the trivial example above, it looks like:

```
@timeout_decorator.timeout(1, use_signals=False)
def fn(i):
    time.sleep(2)
    return i

def wrapped_fn(i):
  try:
    return fn(i)
  except timeout_decorator.TimeoutError:
    return None

with concurrent.futures.ProcessPoolExecutor() as executor:
    for result in executor.map(wrapped_fn, range(100)):
       print(result)
```

You can _almost_ avoid `timeout_decorator` with Py 3's native Futures, [which have timeouts](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future), but those timeouts won't lead to the stalled process being killed, so if the timeouts are frequent enough, your entire process pool could stall.

I realize it could be a little rude to point to using an alternate library here, and that isn't my intention—we received a _ton_ of value from Pebble and very much appreciate noxdafox's work. But we did find that the landscape looked different under Python 3 than it had last time we looked (under Python 2) and wanted to share where we landed in case others run into this same issue! (And maybe seeing an alternate implementation will eventually help Pebble discover how to fix the issue)

**Comment by noxdafox** [comment]
Sorry for the late reply.

What reported is expected behaviour. The `BrokenPoolError` is a catastrophic event which gets raised only when the Pool logic reaches a state in which it cannot continue to operate correctly. It is borrowed from [`concurrent.futures.BrokenExecutor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.BrokenExecutor).

In such cases, the Pool logic terminates quickly releasing all its resources. The pending futures will not be released and the rationale for that was to avoid further complicating the logic when handling a problematic condition.

Depending on your service logging, you should see the core issue in the service output as these internal errors are actually printed out. You might need to trace back some time due to the asynchronous nature of the problem. From my experience, most of the `BrokenPoolErrors` are caused by OOM.

I will see if I can release the pending futures setting a meaningful exception but I cannot ensure this is feasible. The recommendation is to always handle these issues by cancelling all pending futures in your service.

**Comment by noxdafox** [comment]
Had some time to look more closely to this issue.

We cannot cancel pending futures from within the `ProcessPool` logic as this is a decision which belongs to the User and not the Pool itself. 

Nevertheless, the behaviour observed above is indeed incorrect as the pool shall not deadlock in such case but rather raise `TimeoutError` during the iteration over the results. This was not happening due to an incorrect handling of timeouts in such corner cases (a broken pool).

I fixed this issue and now the above logic will raise `TimeoutError` as expected leaving the pool User to decide what to do with the pending futures.

 ```python
with ProcessPool() as pool:
    future = pool.map(fn, range(100), timeout=5)
    iterator = future.result()
    pool._context.state = ERROR

    while True:
        try:
            result = next(iterator)
        except TimeoutError:
            if not pool.active:
                print("Error: pool is not active")
                future.cancel()
                break
        except StopIteration:
            break
        except Exception as err:
            print(err)
```

**Comment by noxdafox** [comment]
Fix added in [release 4.6.0](https://github.com/noxdafox/pebble/releases/tag/4.6.0).

Closing this issue, please re-open it if the problem persists.
